### PR TITLE
add typesafe `std/paths`, `std/files`, `std/dirs`, `std/symlinks`

### LINKED ISSUE #54: New path handling module that uses Nim's type system, using Path instead of string
@Araq in https://github.com/nim-lang/Nim/issues/8268#issuecomment-405817765 suggested:
> You know... I think we need type Path = distinct string; type Filename = distinct string; type FileExt = distinct string and a new path handling module that uses Nim's type system.

let me expand on this a bit with other ideas to open a discussion

## example usage
```nim
import ospaths2

let path ="/tmp//foo.txt".Path # cross platform, works on posix, windows, etc (no allocation)
when defined(posix)
  doAssert path.internal == "/tmp//foo.txt" # just returns internal representation (no allocation)
  doAssert $path == "/tmp/foo.txt" # note: `//` got normalized to `/`
when defined(Windows)
  doAssert path.internal == "/tmp/foo.txt"
  doAssert $path == r"C:\tmp\foo.txt"
  doAssert path == r"C:\TMP\\foo.txt" # platform specific path compare (note the case insensitivity and \\

# revamp functions that accept paths to be type-safe, self-documenting, and avoid confusion between string params (eg file contents) and path params
copyFile($"/tmp/foo.txt", $"/tmp/foo2.txt") # ok
copyFile("/tmp/foo.txt", "/tmp/foo2.txt") # ok via implicit conversion
```

## related work
* C++ boost: https://theboostcpplibraries.com/boost.filesystem-paths
* super relevant discussion from D forum: [Path as an object in std.path](https://forum.dlang.org/thread/dnfvkrnbpxzqmwcbjgfy@forum.dlang.org?page=1) ; contains lots of pros / cons of such approach in the thread

## benefits
* more type safety: eg, we can have `==` overloaded to be platform specific path comparison 
* self-documenting APIs, better encapsulation
* avoid confusion between string params (eg file contents) and path params
* cross platform paths, eg this could work on windows: "/tmp/foo.txt".Path

## cons
see arguments from Walter Bright here: https://forum.dlang.org/post/koqa9h$2l7j$1@digitalmars.com
- cannot hope to duplicate the rich interface available for strings
- APIs that deal with filenames take strings and return strings, not Path objects. Your code gets littered with path and filename components that are sometimes Paths and sometimes strings and sometimes both
- People like writing paths as "/etc/hosts", not Path("/etc/hosts"). People will not stand for a Path constructor that winds up allocating memory so it can rewrite the string in a canonical path representation.
- There really isn't any such thing as a portable path representation. It's more than just \ vs /. There are the drive prefixes in Windows that have no analog in Linux. Sometimes case matters in Linux, where it would be ignored under Windows. There are 8.3 issues sometimes. The only thing you can do is come up with a subset of what works across systems, and then of course you have to go back to using strings when you need to access D:\foo\abc.c

## question
* how easy would be to migrate code? automated tooling possible
* estimate code breakage; can breakage be 100% avoided? (as always, problem is third party libs)
* efficiency: will that lead to more or less efficient code? eg possibly less conversion in most apps assuming most string handling happens early on

## design decisions
* `type Path = distinct string` or `type Path = object ...` ? the latter allows more efficient operations potentially
* should `$path` be `path.internal` or `path.canonical` ?
* should Path implicitly convert to string? cf in D we can use `alias this` to have that; what about Nim?
if it doesn't implicitly convert to string, a ton of code will break. => SUGGESTION: should implicitly convert
* which operations allocate?
```nim
var myPath = "/etc/hosts".Path # does this allocate? I think it should not (eg in case it's unused in some code path)
let a = $myPath # this should allocate (or: this always allocates)
let ok = myPath == "/ETC//hosts" # does this allocate?
```
* should we allow this on windows: `let path ="/tmp/foo.txt".Path`
* should we allow this on windows: `let path =r"C:\tmp\foo.txt".Path`

## links
* Relevant: https://irclogs.nim-lang.org/06-09-2018.html#14:17:36

* @mratsim
> Maybe we should have a Path "distinct string" type that stores validated canonical paths.

### LINKED ISSUE #71: [compiler/pathutils] `AbsoluteFile,AbsoluteDir,RelativeFile,RelativeDir` is too complicated
now that there's a talk of starting to integrate compiler/pathutils inside ospaths to allow reusing its `relativeTo` in `ospaths` (as suggested here: https://github.com/nim-lang/Nim/pull/8166#issuecomment-432796412) it's time to look into design decisions more carefully.
I actually had a WIP PR that moved things around to allow for that (see branch https://github.com/timotheecour/Nim/tree/pr_ospaths_pathutils [it's WIP]) but I don't like the resulting API's.

let me denote by `strongly typed paths` the types introduced in compiler/pathutils:
```nim
type
  AbsoluteFile* = distinct string
  AbsoluteDir* = distinct string
  RelativeFile* = distinct string
  RelativeDir* = distinct string
```

Let's contrast this with following alternative, dubbed `simply typed paths`:
```nim
type
  Path* = distinct string
```

IMO, distinguishing (what's more, at compile time) between these 4 variants **doesn't pull its weight**, and results in a lot of procs being duplicated. It's already the case in pathutils, and, if we apply same logic to rest of procs in ospaths, the problem is only gonna get worse.

## problem 1: dealing with paths of unknown type
* suppose an application deals with some paths of unknown origin (paths could be either absolute or relative, files or dirs) where we don't care which category these belong to; there's no simple way to express these in terms of the strongly typed paths, nor to use strongly typed API's ; this **forces user code to be generic, which is "contagious"**, what would've started as non-generic code has to become generic, adding complexity

* another example: `isAbsolute` itself can't be written using strongly typed paths; it has to be written as `isAbsolute(string)`, losing the benefit of type safety + normalization we get by `simply typed paths`

## problem 2: very little benefit from distinguishing between those
eg:

* lots of cases where the path type makes 0 difference:
```
  proc `==`*(x, y: AbsoluteFile): bool = eqImpl(x.string, y.string)
  proc `==`*(x, y: AbsoluteDir): bool = eqImpl(x.string, y.string)
  proc `==`*(x, y: RelativeFile): bool = eqImpl(x.string, y.string)
  proc `==`*(x, y: RelativeDir): bool = eqImpl(x.string, y.string)

proc isEmpty*(x: AbsoluteFile): bool {.inline.} = x.string.len == 0
proc isEmpty*(x: AbsoluteDir): bool {.inline.} = x.string.len == 0
proc isEmpty*(x: RelativeFile): bool {.inline.} = x.string.len == 0
proc isEmpty*(x: RelativeDir): bool {.inline.} = x.string.len == 0
```

likewise with `unixToNativePath`, and many other procs in `ospaths`

* lots of cases where the path type makes little difference:
eg: `joinPaths(path1, path2)` should work with:
(AbsoluteDir | RelativeDir, AbsoluteFile | AbsoluteDir | RelativeFile | RelativeDir)
so we end up with either generics or with very redundant proc overloads; the only advantage of type safety we get here is preventing 1st arg `path1` to be a File; this doesn't seem like a good tradeoff

## problem 3: not clear how to represent / provide API's for symlinks, filenames etc
is a filename (aka basename) just a string? or yet another distinct string? or a RelativeFile ? if it's just a string (the most likely option), we lose the benefits of type safety from hereafter for the resulting basename, eg:
```
let file = RelativeFile("foo1/foo2/bar.txt")
let filename = extractFileName(file)
let dir: RelativeDir = dirName(dirName(file))
let file2 = dir / filename # error: 2nd arg should be a `AbsoluteFile | AbsoluteDir | RelativeFile | RelativeDir`; not a `string`
```

## [EDIT] problem 4: the type safety guarantees are not actual guarantees, and often violated in practice
the guarantees offered by pathutils.AbsoluteFile + friends are rather moot because code like AbsoluteFile(file) makes 0 checking (nor can it since it's an object constructor).
In practice these "guarantees" are often violated, eg see:
* https://github.com/nim-lang/Nim/pull/13235
* https://github.com/nim-lang/Nim/issues/13121
* `AbsoluteFile"stdin"` in modules.nim
* `AbsoluteFile"command line"` in modules.nim
* `AbsoluteFile(x.strVal)` in pragmaLine which is a lie when user uses `{.loc: "file.nim".}`, or uses `--excessiveStackTrace:off`

## alternative A1 (simplest)

* just use a single type:
```nim
type
  Path* = distinct string
```
`Path` could represent a normalized path, it'd be normalized during construction, eg:
```nim
let a=Path"/abc//def/" # calls `normalize` (aka canon)
doAssert a.string == "/abc/def"

# partial type safety
proc write(path: Path, contents: string) # writeFile is implicit; and no possible confusion between path and contents
# still need to distinguish these:
proc removeFile(file: Path) 
proc removeDir(dir: Path) 
```

## alternative A2 (typesafe and flexible)
use a class hierarchy
```nim
type
  Path* = object of RootObj
    s: string
  PathFile* = object of Path
  PathDir* = object of Path

template p(s: string): Path = Path(s: s.normalize)
template `$`(path: Path): string = path.s
let path = p"abc//def/"
echo path # "abc/def

# full type safety:
proc write(path: PathFile, contents: string)
proc remove(file: PathFile) 
proc remove(dir: PathDir) 
```
Note: I'm not distinguishing between absolute vs relative here; IMO it doesn't pull its weight.
API's shall use either Path, PathFile, or PathDir as appropriate in proc definitions.

We still get **benefits of type safety at compile time** where we need to distinguish, as well **ability to use paths of unknown kind** (a path, which can be a file or dir)

One thing to consider is whether this leads to runtime performance issues (due to dynamic dispatch) of practical impact. If it doesn't, this may be the best option.

### PR DESCRIPTION
**Pull request by ringabout**
Ref https://github.com/nim-lang/RFCs/issues/437
fixes https://github.com/nim-lang/RFCs/issues/54
fixes https://github.com/nim-lang/RFCs/issues/71

Now `Path` is defined as `distinct string`, refined functions should be based on this signature.

- [x] documentaion
- [x] tests

**todo** in the following PRs
- [ ] copyFile, copyDir 
- [ ] createFile
- [ ] runnableExamples
- [ ] documentaion links
- [x] `std/appdirs`

### CODE REVIEW

**Comment by juancarlospaco** [comment]
Very interesting work; This will do symlinks/hardlinks too?. 🙂:+1:

**Comment by Araq** [comment]
> Now Path is defined as distinct string, refined functions should be based on this signature.

Agreed. In the past I've argued for even more fine-grained distinctions like AbsolutePath vs AbsoluteFile and Relative variants but I changed my mind. A single Path type is good enough and slightly easier to work with.

**Comment by beef331** [comment]
With the new Os/Io implementation, will the nimscript operations be moved to another module so that we can use things like `fileExists` of the host machine at compile time when targeting an OS that doesnt support it?

**Comment by Araq** [comment]
> With the new Os/Io implementation, will the nimscript operations be moved to another module so that we can use things like fileExists of the host machine at compile time when targeting an OS that doesnt support it?

There are no plans wrt compile-time evaluation for os.nim. Everything we change or add needs an RFC to justify it because it's so easy to get this wrong.

**Comment by ringabout** [comment]
This PR doesn't contain any API addition, which will be added later. `std/files`, `std/dirs` and `std/symlinks` contain minimal APIs.

**Comment by Araq** [comment]
@ringabout So ... is it ready to review?

**Comment by ringabout** [comment]
Yes

File: lib/std/paths.nim [hunk 1]
```diff
@@ -0,0 +1,322 @@
+import std/private/osseps
+export osseps
+
+from std/private/ospaths2 {.all.} import joinPathImpl, joinPath, splitPath,
```
Comment (Araq) [inline lib/std/paths.nim]: Can be done later but `.all` here is not required. Instead export "all" in "private/ospaths2.nim" as it's a private module.

File: lib/std/paths.nim [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                                      changeFileExt, addFileExt, cmpPaths, splitFile,
+                                      unixToNativePath, absolutePath, normalizeExe,
+                                      normalizePath
+export ReadDirEffect, WriteDirEffect
+
+type
+  Path* = distinct string
+
+
+func joinPath*(head, tail: Path): Path {.inline.} =
```
Comment (Araq) [inline lib/std/paths.nim]: Remove this, `/` is good enough.

File: lib/std/paths.nim [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+  ##
+  ## See also:
+  ## * `joinPath(parts: varargs[Path]) proc`_
+  ## * `/ proc`_
+  ## * `splitPath proc`_
+  ## * `uri.combine proc <uri.html#combine,Uri,Uri>`_
+  ## * `uri./ proc <uri.html#/,Uri,string>`_
+  result = Path(joinPath(head.string, tail.string))
+
+func joinPath*(parts: varargs[Path]): Path =
```
Comment (Araq) [inline lib/std/paths.nim]: Instead offer an `add` operation.

File: lib/std/paths.nim (hunk 1 above)
Comment (ringabout) [reply lib/std/paths.nim]: Yeah, but "std/private/ospaths2" is imported and re-exported by `std/os`.

File: lib/std/paths.nim [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+  ## Returns a normalized path for the current OS.
+  ##
+  ## See also:
+  ## * `absolutePath proc`_
+  ## * `normalizePath proc`_ for the in-place version
+  result = Path(normalizedPath(path.string))
+
+proc normalizePathEnd*(path: var Path, trailingSep = false) {.borrow.}
+
+proc normalizePathEnd*(path: Path, trailingSep = false): Path {.borrow.}
```
Comment (Araq) [inline lib/std/paths.nim]: No. Overloading by `var` here repeats the existing design bug! Instead remove the non-var version. There is `dup` for these things.

File: lib/std/paths.nim (hunk 2 above)
Comment (ringabout) [reply lib/std/paths.nim]: Sure.

File: lib/std/paths.nim [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+  ##
+  ## See also:
+  ## * `searchExtPos proc`_
+  ## * `splitFile proc`_
+  ## * `extractFilename proc`_
+  ## * `changeFileExt proc`_
+  ## * `addFileExt proc`_
+  result = Path(lastPathPart(path.string))
+
+func changeFileExt*(filename, ext: Path): Path {.inline.} =
```
Comment (Araq) [inline lib/std/paths.nim]: `ext` should not be of type `Path` here.

File: lib/std/paths.nim [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+  ## See also:
+  ## * `joinPath(head, tail) proc`_
+  ## * `joinPath(parts: varargs[Path]) proc`_
+  ## * `/ proc`_
+  ## * `/../ proc`_
+  ## * `relativePath proc`_
+  let res = splitPath(path.string)
+  result = (Path(res.head), Path(res.tail))
+
+func splitFile*(path: Path): tuple[dir, name, ext: Path] {.inline.} =
```
Comment (Araq) [inline lib/std/paths.nim]: `ext` shall be of type `string`.

File: lib/std/paths.nim (hunk 3 above)
Comment (ringabout) [reply lib/std/paths.nim]: Sure.

File: lib/std/paths.nim [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+func `/../`*(head, tail: Path): Path {.inline.} =
+  ## The same as ``parentDir(head) / tail``, unless there is no parent
+  ## directory. Then ``head / tail`` is performed instead.
+  ##
+  ## See also:
+  ## * `/ proc`_
+  ## * `parentDir proc`_
+  Path(`/../`(head.string, tail.string))
+
+proc searchExtPos*(path: Path): int {.inline.} =
```
Comment (Araq) [inline lib/std/paths.nim]: No need to export this proc.

File: lib/std/paths.nim [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+  ##
+  ## See also:
+  ## * `searchExtPos proc`_
+  ## * `splitFile proc`_
+  ## * `extractFilename proc`_
+  ## * `lastPathPart proc`_
+  ## * `addFileExt proc`_
+  result = Path(changeFileExt(filename.string, ext.string))
+
+func addFileExt*(filename, ext: Path): Path {.inline.} =
```
Comment (Araq) [inline lib/std/paths.nim]: ext should be of type string.

File: lib/std/paths.nim (hunk 1 above)
Comment (Araq) [reply lib/std/paths.nim]: So? Only re-export in std/os what should be re-exported.

File: lib/std/paths.nim (hunk 8 above)
Comment (ringabout) [reply lib/std/paths.nim]: done

File: lib/std/paths.nim (hunk 1 above)
Comment (ringabout) [reply lib/std/paths.nim]: done

**Pull request merged**